### PR TITLE
Improve SqlAlchemy session handling for `QueryBuilder`

### LINKED ISSUE #3372: Use Aldjemy to create the session for Django `QueryBuilder` implementation
Currently we use the one from `aiida.backends.sqlalchemy`, but this will most likely not use the session correctly and will cause problems. We should see if we can use the session managed by aldjemy instead.

In `aiida.orm.implementation.django.querybuilder` replace the `get_session` with something like:
```
def get_session():
    from aldjemy.orm import get_session
    return get_session()
```
This fixes the following test:
```
from aiida.backends.djsite.db.models import DbNode, DbUser
from django.db import transaction
with transaction.atomic():                                
    user = DbUser(email='aiida@localhost2', first_name='t', last_name='t', institution='t')
    user.save()    
    node = DbNode(user_id=user.pk)
    node.save()

    # Loading it through the Django ORM should work, even when in a transaction
    loaded = DbNode.objects.get(pk=node.pk, user_id=user.pk)

    # Try to load it through the QueryBuilder which will fail if the session of SqlAlchemy is not correctly setup
    assert QueryBuilder().append(Node, filters={'id': node.pk}).count() != 0
```

However, the unittests will still fail so the implementation of `get_session` is not enough. Probably some more configuration is necesary

### LINKED ISSUE #3645: Sessions not handled correctly in QB for asynchronous function calls
I have found in my [`aiida-optimade`](https://github.com/aiidateam/aiida-optimade) REST API implementation exposing an AiiDA DB in the OPTiMaDe spec format that I get the [SQLA QueuePool error](https://docs.sqlalchemy.org/en/13/errors.html#queuepool-limit-of-size-x-overflow-y-reached-connection-timed-out-timeout-z) when making more than ~15 HTTP GET requests.

I have remedied this in my implementation by an AiiDA hack that creates a new DB session per request and closes it at the end.

It seems to me the native AiiDA REST API will have a similar problem for too many (concurrent) requests.

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #3645 and fixes #3372

### CODE REVIEW

**Comment by CasperWA** [comment]
Concerning using this in a REST-API server, i.e., `aiida-optimade`, what do you suggest to do?
I have tried to retrieve the session via `get_manager().get_backend().get_session()` before handling a request and then calling `session.close()` at the end, but this still prompts a QueuePool overflow for both backends.

File: aiida/backends/djsite/manager.py [hunk 1]
```diff
@@ -32,6 +32,8 @@ def _load_backend_environment(self):
 
     def reset_backend_environment(self):
         """Reset the backend environment."""
+        from aiida.orm.implementation.django.querybuilder import reset_session
```
Comment (CasperWA) [inline aiida/backends/djsite/manager.py]:
```suggestion
        from . import reset_session
```

File: aiida/backends/sqlalchemy/manager.py [hunk 2]
```diff
@@ -62,7 +62,7 @@ def reset_backend_environment(self):
         from aiida.backends import sqlalchemy
         if sqlalchemy.ENGINE is not None:
             sqlalchemy.ENGINE.dispose()
-        sqlalchemy.SCOPED_SESSION_CLASS = None
+        sqlalchemy.SESSION_FACTORY = None
```
Comment (CasperWA) [inline aiida/backends/sqlalchemy/manager.py]: Shouldn't this method simply call `aiida.backends.sqlalchemy.reset_session()`?

File: aiida/backends/djsite/manager.py (hunk 1 above)
Comment (CasperWA) [reply aiida/backends/djsite/manager.py]: I think this is still an issue?

File: aiida/orm/implementation/querybuilder.py [hunk 3]
```diff
@@ -9,6 +9,12 @@
 ###########################################################################
 """Backend query implementation classes"""
```
Comment (CasperWA) [inline aiida/orm/implementation/querybuilder.py]: Maybe note here that the general QueryBuilder here is based on the assumption that all queries are performed with SQLAlchemy, no matter the actual backend? I.e., just copy some stuff in here from your commit messages.

File: aiida/orm/implementation/sqlalchemy/backend.py [hunk 4]
```diff
@@ -101,6 +98,18 @@ def transaction():
                 # Make sure to commit the outermost session
                 session.commit()
 
+    @staticmethod
+    def get_session():
+        """Return a database session that can be used by the `QueryBuilder` to perform its query.
+
+        If there is an exception within the context then the changes will be rolled back and the state will
```
Comment (CasperWA) [inline aiida/orm/implementation/sqlalchemy/backend.py]: Shouldn't this be under the description of the `transaction` method?

File: aiida/backends/sqlalchemy/manager.py (hunk 2 above)
Comment (sphuber) [reply aiida/backends/sqlalchemy/manager.py]: Sure does :)

**Pull request merged**